pkg/scheduler/schedule_one.go

```diff
@@ -1074,9 +1074,18 @@ func (sched *Scheduler) handleSchedulingFailure(ctx context.Context, fwk framewo
 			// ignore this err since apiserver doesn't properly validate affinity terms
 			// and we can't fix the validation for backwards compatibility.
 			podInfo.PodInfo, _ = framework.NewPodInfo(cachedPod.DeepCopy())
-			if err := sched.SchedulingQueue.AddUnschedulableIfNotPresent(logger, podInfo, sched.SchedulingQueue.SchedulingCycle()); err != nil {
-				logger.Error(err, "Error occurred")
-			}
+			schedulingCycle := sched.SchedulingQueue.SchedulingCycle()
+			go func() {
+				// When handling events takes time in the queue, AddUnschedulableIfNotPresent could be blocked by a shared lock within the queue,
+				// which negatively impacts the scheduling throughput.
+				// So, we put an unschedulable pod back to the queue in a goroutine.
+				//
+				// We handle a nominatingInfo outside goroutine below,
+				// and hence there shouldn't be any race condition etc that caused by doing this in goroutine.
+				if err := sched.SchedulingQueue.AddUnschedulableIfNotPresent(logger, podInfo, schedulingCycle); err != nil {
+					logger.Error(err, "Error occurred when putting an unschedulable pod back to the queue", "pod", klog.KObj(pod))
+				}
+			}()
```

Comment:
> what will happen if we call AddNominatedPod first, I think this is the most case because starting goroutine is slow.

The order doesn't matter; whether it's called before or after `AddUnschedulableIfNotPresent`, nothing wrong happens.

> what about the inFlightEvents, because we may call Done first in handleSchedulingFailure.

We don't call `Done` in `handleSchedulingFailure` side because of `calledDone` at L1089.

Reply:
> We don't call Done in handleSchedulingFailure side because of calledDone at L1089.

Yes, misread the defer() function.

Reply:
If we are calling Done in the goroutine, and the scheduler keeps trying to schedule more pods, it means that more pods could be now stay in "inflight" at a time, which could increase memory consumption.

It sounds like we don't currently have evidence that supports that this change is needed. It only has the potential to improve scheduling when there are many unschedulable pods. Those cases have been historically slow and we recommend our users not to get into that state. For example, we don't recommend using fake node selectors to prevent pods from scheduling. That's why we introduced scheduling gates as a more efficient way of doing this.

So, I'm not too convinced that this is the right time to merge this change.

Reply:
> more pods could be now stay in "inflight" at a time, which could increase memory consumption.

Fair point. That's definitely a risk.
Okay, let's close this PR for now, just keep it in mind, and revisit here if we still find it necessary after all the other improvements around (https://github.com/kubernetes/kubernetes/pull/125692, https://github.com/kubernetes/kubernetes/pull/125820, https://github.com/kubernetes/kubernetes/issues/120622) are done.

/close